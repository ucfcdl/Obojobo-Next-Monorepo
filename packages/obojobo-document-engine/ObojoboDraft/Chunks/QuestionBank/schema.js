import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

import ParameterNode from '../../../src/scripts/oboeditor/components/parameter-node'

const QUESTION_BANK_NODE = 'ObojoboDraft.Chunks.QuestionBank'
const SETTINGS_NODE = 'ObojoboDraft.Chunks.QuestionBank.Settings'
const QUESTION_NODE = 'ObojoboDraft.Chunks.Question'

const SELECT_TYPES = ['sequential', 'random', 'random-unseen']

const schema = {
	blocks: {
		'ObojoboDraft.Chunks.QuestionBank': {
			nodes: [
				{ match: [{ type: SETTINGS_NODE }], min: 1, max: 1 },
				{ match: [{ type: QUESTION_NODE }, { type: QUESTION_BANK_NODE }], min: 1 }
			],
			normalize: (change, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_REQUIRED: {
						if (index === 0) {
							const block = Block.create({
								type: SETTINGS_NODE
							})
							return change.insertNodeByKey(node.key, index, block)
						}
						const block = Block.create({
							type: QUESTION_NODE
						})
						return change.insertNodeByKey(node.key, index, block)
					}
					case CHILD_TYPE_INVALID: {
						if (index === 0) {
							const block = Block.create({
								type: SETTINGS_NODE
							})
							return change.wrapBlockByKey(child.key, block)
						}
						return change.wrapBlockByKey(child.key, {
							type: QUESTION_NODE
						})
					}
				}
			}
		},
		'ObojoboDraft.Chunks.QuestionBank.Settings': {
			nodes: [
				{
					match: [
						{
							type: 'Parameter'
						}
					],
					min: 2,
					max: 2
				}
			],
			normalize: (change, error) => {
				const { node, child, index } = error
				switch (error.code) {
					case CHILD_REQUIRED: {
						if (index === 0) {
							const block = Block.create(
								ParameterNode.helpers.oboToSlate({
									name: 'choose',
									value: 'Infinity',
									display: 'Choose'
								})
							)
							return change.insertNodeByKey(node.key, index, block)
						}
						const block = Block.create(
							ParameterNode.helpers.oboToSlate({
								name: 'select',
								value: 'sequential',
								display: 'Select',
								options: SELECT_TYPES
							})
						)
						return change.insertNodeByKey(node.key, index, block)
					}
					case CHILD_TYPE_INVALID: {
						return change.withoutNormalization(c => {
							c.removeNodeByKey(child.key)
							if (index === 0) {
								const block = Block.create(
									ParameterNode.helpers.oboToSlate({
										name: 'choose',
										value: 'Infinity',
										display: 'Choose'
									})
								)
								return c.insertNodeByKey(node.key, index, block)
							}
							const block = Block.create(
								ParameterNode.helpers.oboToSlate({
									name: 'select',
									value: 'sequential',
									display: 'Select',
									options: SELECT_TYPES
								})
							)
							return c.insertNodeByKey(node.key, index, block)
						})
					}
				}
			}
		}
	}
}

export default schema
