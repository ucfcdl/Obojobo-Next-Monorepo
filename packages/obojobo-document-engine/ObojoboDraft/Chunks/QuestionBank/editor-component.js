import React from 'react'
import { Block } from 'slate'
import Common from 'Common'

import emptyQB from './empty-node.json'
import emptyQuestion from '../Question/empty-node.json'

const { Button } = Common.components

class QuestionBank extends React.Component {
	constructor(props) {
		super(props)
		this.state = props.node.data.get('content')
	}
	delete() {
		const editor = this.props.editor
		return editor.removeNodeByKey(this.props.node.key)
	}
	addQuestion() {
		const editor = this.props.editor
		const newQuestion = Block.create(emptyQuestion)
		return editor.insertNodeByKey(this.props.node.key, this.props.node.nodes.size, newQuestion)
	}
	addQuestionBank() {
		const editor = this.props.editor

		const newQuestion = Block.create(emptyQB)

		return editor.insertNodeByKey(this.props.node.key, this.props.node.nodes.size, newQuestion)
	}
	render() {
		return (
			<div className={'obojobo-draft--chunks--question-bank editor-bank'}>
				<button className="editor--page-editor--delete-node-button" onClick={() => this.delete()}>
					X
				</button>
				{this.props.children}
				<Button className={'buffer'} onClick={() => this.addQuestion()}>
					{'Add Question'}
				</Button>
				<Button className={'buffer'} onClick={() => this.addQuestionBank()}>
					{'Add Question Bank'}
				</Button>
			</div>
		)
	}
}

export default QuestionBank
