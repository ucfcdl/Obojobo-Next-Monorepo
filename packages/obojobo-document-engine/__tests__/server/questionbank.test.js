/* eslint no-undefined: 0 */

import QuestionBank from '../../server/questionbank'

describe('QuestionBank', () => {
	const questionBank = new QuestionBank({}, { id: 'mockQuestionBank' }, {})

	test('registers expected events', () => {
		expect(questionBank.registerEvents).toHaveBeenCalledTimes(1)
		const events = questionBank.registerEvents.mock.calls[0][0]
		expect(questionBank.registerEvents.mock.calls[0]).toMatchSnapshot()
		expect(events['ObojoboDraft.Sections.Assessment:sendToClient']).toBe(
			questionBank.onSendToClient
		)
	})

	test('removes children on sendToClient', () => {
		questionBank.children = 'mockChildren'
		questionBank.onSendToClient()
		expect(questionBank.children).toEqual([])
	})
})
