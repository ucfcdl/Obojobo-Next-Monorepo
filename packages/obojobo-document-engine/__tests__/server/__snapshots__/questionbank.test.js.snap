// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionBank registers expected events 1`] = `
Array [
  Object {
    "ObojoboDraft.Sections.Assessment:sendToClient": [Function],
  },
]
`;
